pkg/resource/deploy/builtins.go

```diff
@@ -113,65 +113,126 @@ func (p *builtinProvider) Configure(context.Context, plugin.ConfigureRequest) (p
 	return plugin.ConfigureResponse{}, nil
 }
 
-const stackReferenceType = "pulumi:pulumi:StackReference"
+const (
+	stackReferenceType = "pulumi:pulumi:StackReference"
+	stashType          = "pulumi:index:Stash"
+)
 
 func (p *builtinProvider) Check(_ context.Context, req plugin.CheckRequest) (plugin.CheckResponse, error) {
 	typ := req.URN.Type()
-	if typ != stackReferenceType {
+	if typ != stackReferenceType && typ != stashType {
 		return plugin.CheckResponse{}, fmt.Errorf("unrecognized resource type '%v'", typ)
 	}
 
-	// We only need to warn about this in Check. This won't be called for Reads but Creates or Updates will
-	// call Check first.
-	msg := "The \"pulumi:pulumi:StackReference\" resource type is deprecated. " +
-		"Update your SDK or if already up to date raise an issue at https://github.com/pulumi/pulumi/issues."
-	p.diag.Warningf(diag.Message(req.URN, msg))
+	if typ == stackReferenceType {
```

Comment:
Optional style nit: This would be easier to grok if it was a switch statement:

```go
switch typ {
case stackReferenceType:
	...
case stashType:
	...
default:
	contract.Failf("expected resource type %q or %q, got %q", stackReferenceType, stackType, typ)
}
```